ENH: show percentiles in timestamp describe (#30164)

david-cortes opened the pull request:
closes https://github.com/pandas-dev/pandas/issues/30164

When creating a pandas Series of timestamps and calling `describe` on it, it will not show the percentiles of the data, even if these are specified as arguments (`s.describe(percentiles = [0.25, 0.5, 0.75])`).

This PR solves the issue by introducing a new describe logic for timestamps that would:
* Add the percentiles in the same way as for numeric.
* Rename `first` and `last` to `min` and `max` in order to match numeric types.
* Add the mean of the series.
* ~~Show datetime columns alongside numerics by default in the describe method for DataFrames.~~ (just realized there were other issues in which this was deemed to be the desirable behavior)

After this, it more or less matches the functionality of R's `summary` on `POSIXct` and `Date` types.

Why I think this is a good idea: I oftentimes find myself trying to inspect tables of mixed types, and want to get a quick overview of which kind of ranges and variations are there in each column, including timestamps. Currently this is very inconvenient in pandas, and they are not displayed alongside numeric columns by default and don't match the names if passing `include = "all"`, which is what I want to see when I call `DataFrame.describe`.

Examples:
```python
import numpy as np, pandas as pd
dt_list = [
    "2019-01-01 00:01:00",
    "2019-01-01 00:02:00",
    "2019-01-01 00:02:10",
    "2019-01-01 00:10:01",
    "2019-01-01 00:11:00"
]
s = pd.to_datetime(pd.Series(dt_list))
s.describe()
```
```python
pd.DataFrame({
    "col1" : s,
    "col2" : np.arange(s.shape[0])
}).describe(include = "all")
```

Update: corrected the coding style to comply with `black` and pass the automatic test.

gfyoung commented on pandas/core/generic.py:
```diff
@@ -9806,11 +9788,28 @@ def describe_categorical_1d(data):
             return pd.Series(result, index=names, name=data.name, dtype=dtype)
 
+        def describe_timestamp_1d(data):
+            tz = data.dt.tz
```
Reference issue number above this.

pep8speaks commented:
Hello @david-cortes! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/core/generic.py`](https://github.com/pandas-dev/pandas/blob/c72764a618d96da608201d45f0517be370098cdd/pandas/core/generic.py):

> [Line 9792:13](https://github.com/pandas-dev/pandas/blob/c72764a618d96da608201d45f0517be370098cdd/pandas/core/generic.py#L9792): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '

* In the file [`pandas/tests/frame/test_analytics.py`](https://github.com/pandas-dev/pandas/blob/c72764a618d96da608201d45f0517be370098cdd/pandas/tests/frame/test_analytics.py):

> [Line 784:53](https://github.com/pandas-dev/pandas/blob/c72764a618d96da608201d45f0517be370098cdd/pandas/tests/frame/test_analytics.py#L784): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 796:77](https://github.com/pandas-dev/pandas/blob/c72764a618d96da608201d45f0517be370098cdd/pandas/tests/frame/test_analytics.py#L796): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

david-cortes commented:
Added the issue number in the code, and also modified it to handle cases with empty inputs. Modified the tests on `describe` in `Series` and `DataFrame` too.

MarcoGorelli commented:
Hi @david-cortes - seems like one of the Azure jobs is failing (but the build isn't there anymore so we can't check it), and that there are now some conflicts with _master_. Is this something you're still working on?

david-cortes committed:
ENH: show percentiles in timestamp describe (#30164)

david-cortes commented:
@MarcoGorelli : No, not working on anything else here, the PR is finished. Solved the merge conflicts now (they were caused by some tests moving to separate files).

From what I get of the automatic checks that were failing, they were caused by something unrelated to this PR.

jreback commented on pandas/core/generic.py:
```diff
@@ -9656,11 +9638,30 @@ def describe_categorical_1d(data):
+        def describe_timestamp_1d(data):
+            # GH-30164
+            tz = data.dt.tz
+            asint = data.dropna().values.view("i8")
```
you don't need the .values.view, just use` len(data.dropna())` for the shape check

data.min() and so on all work and return NaT correctly, none of these gymnastics are needed

jreback replied:
I know that's what the current code does, but let's simplify

jreback reviewed (changes requested):
@david-cortes can you create a new issue to move all of the describe helpers to pandas/io/alogorithms.py (or really we should just create pandas/io/algos/describe.py) and then move everything else.

can do before or after this PR

jreback reviewed (changes requested):
this also needs a whats note for 1.1, put in enhancements

david-cortes committed:
use internal series methods for datetime describe (#30209)

david-cortes commented:
@jreback : Did you mean to say to move them to `pandas/core/algorithms.py`? Would seem out of place in `io`.

jreback commented:
> @jreback : Did you mean to say to move them to `pandas/core/algorithms.py`? Would seem out of place in `io`.

yes though i’d like to create

pandas/core/algos/describe.py

and split up algorithms (so this can. e a step there)

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -18,6 +18,8 @@ Enhancements
 Other enhancements
 ^^^^^^^^^^^^^^^^^^
 
+- :meth:`Series.describe` will now show distribution percentiles for ``datetime`` dtypes, statistics ``first`` and ``last``
```
can you move to api changes section.

jreback commented on pandas/tests/frame/methods/test_describe.py:
```diff
@@ -253,52 +253,19 @@ def test_describe_tz_values(self, tz_naive_fixture):
-                    3,
-                    4,
-                ],
+                "s1": [5, 2, 0, 1, 2, 3, 4, 1.581139],
```
do we have sufficient tests for timedeltas?

can you create a test for Period (which likely don't work), if you'd just xfail it (alt if you'd create an issue)

jreback reviewed (changes requested):
lgtm. small comment ping on green.

david-cortes committed:
more tests for datetimes describe

david-cortes commented:
Moved the changes in the what’s new and added tests for `Period` and `Timedelta`. The describe for `Timedelta` is already handled as a numeric dtype, and the describe for `Period` as object dtype, so those are not affected by this PR.

jreback commented on pandas/tests/series/methods/test_describe.py:
```diff
@@ -29,6 +29,36 @@ def test_describe(self):
         )
         tm.assert_series_equal(result, expected)
 
+        s = Series(
```
ok for now, in a followon we want to parameterize these.

jreback commented:
thanks @david-cortes 

if you woudn't mind parameterize / separating out the other dtypes tests would be great (could be combined with moving the code out from pandas/core/algorithms.py
